src/proto/sim.rs

```diff
@@ -0,0 +1,1103 @@
+                    agg.duration += round.duration;
+                    agg.missed += round.missed;
+                    (len + 1., agg)
+                });
+        avg.rmr /= len;
+        avg.ldh /= len;
+        avg.missed /= len;
+        avg.duration /= len as u32;
+        avg
+    }
+
+    pub fn min<'a>(rounds: impl IntoIterator<Item = &'a RoundStats>) -> RoundStats {
+        rounds
+            .into_iter()
+            .fold(RoundStats::new_max(), |mut agg, round| {
+                agg.rmr = agg.rmr.min(round.rmr);
+                agg.ldh = agg.ldh.min(round.ldh);
+                agg.duration = agg.duration.min(round.duration);
+                agg.missed = agg.missed.min(round.missed);
+                agg
+            })
+    }
+
+    pub fn max<'a>(rounds: impl IntoIterator<Item = &'a RoundStats>) -> RoundStats {
+        rounds
+            .into_iter()
+            .fold(RoundStats::default(), |mut agg, round| {
+                agg.rmr = agg.rmr.max(round.rmr);
+                agg.ldh = agg.ldh.max(round.ldh);
+                agg.duration = agg.duration.max(round.duration);
+                agg.missed = agg.missed.max(round.missed);
+                agg
+            })
+    }
+
+    pub fn avg(rounds: &[RoundStats]) -> RoundStatsAvg {
+        let min = Self::min(rounds);
+        let max = Self::max(rounds);
+        let mean = Self::mean(rounds);
+        RoundStatsAvg { min, max, mean }
+    }
+
+    pub fn diff(&self, other: &Self) -> RoundStatsDiff {
+        RoundStatsDiff {
+            duration: diff_percent(self.duration.as_secs_f32(), other.duration.as_secs_f32()),
+            rmr: diff_percent(self.rmr, other.rmr),
+            ldh: diff_percent(self.ldh, other.ldh),
+            missed: diff_percent(self.missed, other.missed),
+        }
+    }
+}
+
+fn diff_percent(a: f32, b: f32) -> f32 {
+    if a == 0.0 && b == 0.0 {
+        0.0
+    } else if b == 0.0 {
+        -1.0
+    } else if a == 0.0 {
+        1.0
+    } else {
+        (b - a) / a
+    }
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct RoundStatsAvg {
+    pub min: RoundStats,
+    pub max: RoundStats,
+    pub mean: RoundStats,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct RoundStatsAvgDiff {
+    pub min: RoundStatsDiff,
+    pub max: RoundStatsDiff,
+    pub mean: RoundStatsDiff,
+}
+
+impl RoundStatsAvg {
+    pub fn diff(&self, other: &Self) -> RoundStatsAvgDiff {
+        RoundStatsAvgDiff {
+            min: self.min.diff(&other.min),
+            max: self.max.diff(&other.max),
+            mean: self.mean.diff(&other.mean),
+        }
+    }
+
+    pub fn avg(rows: &[Self]) -> Self {
+        let min = RoundStats::min(rows.iter().map(|x| &x.min));
+        let max = RoundStats::min(rows.iter().map(|x| &x.max));
+        let mean = RoundStats::min(rows.iter().map(|x| &x.mean));
+        Self { min, max, mean }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct NetworkStats {
```

Comment:
my understanding is that all these are histograms? could you please add docs to this?